### ISSUE TITLE
WebSocker-Next - Refresh OIDC AccessToken without reconnection

### ISSUE THREAD

**Issue by dometec** [quarkusio/quarkus#43434]
### Description

At the moment, when an access token expires, the WebSocker-Next server closes the connection.
Sometimes, it's a problem since a token life span can be very short (a few minutes) and reconnection can be expensive.
Is it possible to think of some mechanism that allows the client to refresh the token before it expires and not close the websocket connection?

### Implementation ideas

Maybe a PATCH request to the same WebSocket endpoint to update the access token can be a solution?

**Comment by sberyozkin** [comment]
The tricky part is to have an expired but earlier verified token replaced with a token which has also been verified by Quarkus.

**Comment by cescoffier** [comment]
@sberyozkin @mkouba This is an interesting question. Currently, our security architecture is made to be managed during the HTTP upgrade.  Would it be possible to have `@Authenticated` on some methods to check and renew the token if needed? The main issue is that the token cannot be stored in the HTTP request but would have to be in the scope of the session.

**Comment by mkouba** [comment]
> The main issue is that the token cannot be stored in the HTTP request but would have to be in the scope of the session.

And where's the token stored right now? I assume that it's part of the `SecurityIdentity` that is captured during the HTTP upgrade. So from WS next POV we would have to "copy" the refresh logic from the `quarkus-oidc` and replace the captured `SecurityIdentity`. Am I right? CC @michalvavrik

**Comment by michalvavrik** [comment]
@mkouba I don't think so, no. What needs to be done is to expose endpoint, probably non-application one that accepts connection id as a parameter. Then you need to require authentication (you remember that in past, you added a handler that is now removed, that route  handler simply ensured that SecurityIdentity is available), ensure the identity is same (not anonymous, principal matches etc.) use the connection id to switch the identity in `io.quarkus.websockets.next.runtime.SecuritySupport`. We also need to keep in mind that `RoutingContext` will hold the old identity so let's switch it or delete it completely from there, that's not an issue as in WS Next, you can't rely on the Rc.

IMHO it's actually very nice task when there is an agreement how to do it. I like the task, but for this week already have 2 tasks on Quarkus and I still owe you Micrometer tests / moving :-)

**Comment by michalvavrik** [comment]
btw I was referring to this suggestion:

> Maybe a PATCH request to the same WebSocket endpoint to update the access token can be a solution?

if you want the OIDC do it, then it will be more complex and requires more discussion

**Comment by michalvavrik** [comment]
Also it will require authorization, probably need to think about it much more :-)

**Comment by mkouba** [comment]
> I don't think so, no. What needs to be done is to expose endpoint, probably non-application one that accepts connection id as a parameter. Then you need to require authentication (you remember that in past, you added a handler that is now removed, that route handler simply ensured that SecurityIdentity is available), ensure the identity is same (not anonymous, principal matches etc.) use the connection id to switch the identity in `io.quarkus.websockets.next.runtime.SecuritySupport`. We also need to keep in mind that `RoutingContext` will hold the old identity so let's switch it or delete it completely from there, that's not an issue as in WS Next, you can't rely on the Rc.

Ah ok, sorry for confusion. I was thinking about something along the lines of the "auto-refresh" feature on the server (I read the OIDC docs and it seems that we support like that in Quarkus) and not about explicit security identity refresh initiated from the client.

In theory, we don't need a new endpoint for this stuff. Instead, we can reuse the handler used to perform the upgrade and handle HTTP `PATCH` method specifically. There's one problem though - for non WS next client you'll have to pass the connection id to the connected clients somehow (a message sent from an `@OnOpen` callback or something like that).

**Comment by michalvavrik** [comment]
>  I was thinking about something along the lines of the "auto-refresh" feature on the server (I read the OIDC docs and it seems that we support like that in Quarkus) and not about explicit security identity refresh initiated from the client.

I agree we should try to autorefresh tokens when possible. I didn't remember we talked about it before when I wrote my comments. I think we will need new lightweight extension that relies on both OIDC and WebSockets Next. I'll come back to this in a week or two, maybe @sberyozkin will find a time to comment in the meanwhile. I remember Sergey was very interested in this subject.

**Comment by michalvavrik** [comment]
I have played with it. You can initiate authorization code flow and once you have the authorization code, you can pass query params from the redirect URL to the WS JS client and the code can be retrieved during the initial handshake request. But I don't think it is possible to pass state cookie that is required by Quarkus OIDC. Also, same-origin policy is not applied on the WS JS, so I think we would need additional measures to mitigate CSRF. Service type application won't receive refresh token (even if we had client credentials).

I still think that async reauthentication to non-application endpoint is the most trivial way to do this, but we definitely need some expert on OIDC to have a say because I could be missing a lot.

**Comment by michalvavrik** [comment]
We talked about this on one of security meetings and also I checked with @mkouba , I think we should start with something minimalist, provide an interface that will allow users to refresh identity based on a new access token. And document how users can do that, either use wrapper for their messages or hide this in a dekoder etc. (will see how to document it..), but let's start with giving option and if there will be demand, we can provide our own subprotocol later.
